File: backend/app/agents/push_planner_agent.py
from __future__ import annotations

from datetime import UTC, datetime, timedelta

from app.agents.agentscope_runtime import run_structured_agent_sync
from app.agents.common import infer_priority, merge_channels, recommend_channels
from app.agents.context import AgentExecutionContext
from app.agents.structured_models import PushPlanStructuredOutput


def _build_push_plan_user_prompt(
    context: AgentExecutionContext,
    hotspot_result: dict[str, object],
    classification_result: dict[str, object],
    summary_result: dict[str, object],
) -> str:
    news = context.news
    return "\n".join(
        [
            "请根据下面这条 AI 新闻和分析结果给出推送计划，只输出结构化结果，不要额外解释。",
            f"标题：{summary_result.get('title') or news.title}",
            f"来源：{news.source}",
            f"分类：{classification_result.get('category') or news.category or '未分类'}",
            f"热度分值：{hotspot_result.get('score') or news.hot_score}",
            f"热点判断：{hotspot_result.get('is_hot')}",
            f"摘要：{summary_result.get('summary') or news.summary or '无'}",
            f"建议渠道：{', '.join(context.config.get('push_channels') or []) or '无'}",
        ]
    )


def run_push_planner_agent(
    context: AgentExecutionContext,
    hotspot_result: dict[str, object],
    classification_result: dict[str, object],
    summary_result: dict[str, object],
) -> dict[str, object]:
    runtime = context.runtime
    if runtime is not None:
        agent_result = run_structured_agent_sync(
            runtime=runtime,
            agent_name="PushPlannerAgent",
            system_prompt=context.prompt_bundle.get("agent_push_planner_prompt", ""),
            user_prompt=_build_push_plan_user_prompt(context, hotspot_result, classification_result, summary_result),
            structured_model=PushPlanStructuredOutput,
        )
        if agent_result is not None:
            priority = str(agent_result.get("priority") or infer_priority(int(hotspot_result.get("score") or context.news.hot_score or 0)))
            agent_result["channels"] = merge_channels(agent_result.get("channels") or [], recommend_channels(priority))
            return {
                **agent_result,
                "model_name": context.model_name,
                "prompt_version": context.prompt_version,
            }

    score = int(hotspot_result.get("score") or context.news.hot_score or 0)
    category = str(classification_result.get("category") or context.news.category or "行业动态")
    is_hot = bool(hotspot_result.get("is_hot"))
    push_now = is_hot and score >= max(int(context.config["hot_threshold"]), 35) + 10
    if category in {"AI Agent", "大模型", "开源生态"} and score >= max(int(context.config["hot_threshold"]), 35):
        push_now = True

    priority = infer_priority(score)
    push_type = "IMMEDIATE" if push_now else "DIGEST"
    channels = merge_channels(context.config.get("push_channels") or [], recommend_channels(priority))
    planned_at = datetime.now(UTC) if push_now else datetime.now(UTC) + timedelta(hours=6)

    return {
        "push_now": push_now,
        "priority": priority,
        "push_type": push_type,
        "channels": channels,
        "planned_at": planned_at.isoformat(),
        "reason": f"热度评分 {score}，分类 {category}，推送策略 {push_type}",
        "model_name": context.model_name,
        "prompt_version": context.prompt_version,
        "title": summary_result.get("title") or context.news.title,
    }

File: backend-ts/src/agents/push-planner.agent.ts
/**
 * PushPlannerAgent - Push planning agent
 * Outputs push decisions: whether to push, priority, channels, timing
 *
 * Pattern: AI-first rule-fallback
 */

import { Injectable, Logger } from '@nestjs/common';
import { z } from 'zod';

import { AgentScopeRuntimeService } from './agents.runtime';
import type {
  AgentExecutionContext,
  ClassificationOutput,
  EnrichmentOutput,
  HotspotOutput,
  PushPlanOutput,
  SummaryOutput,
} from './agents.types';
import {
  buildModelLabel,
  inferPriority,
  mergeChannels,
  recommendChannels,
} from './agents.common';

// ===== Zod Schema =====

export const PushPlanSchema = z.object({
  push_now: z.boolean(),
  priority: z.string(),
  push_type: z.string(),
  channels: z.array(z.string()).default([]),
  planned_at: z.string(),
  reason: z.string(),
  title: z.string(),
});

@Injectable()
export class PushPlannerAgent {
  private readonly logger = new Logger(PushPlannerAgent.name);

  constructor(private readonly runtime: AgentScopeRuntimeService) {}

  private buildUserPrompt(
    ctx: AgentExecutionContext,
    hotspotResult: HotspotOutput,
    classificationResult: ClassificationOutput,
    summaryResult: SummaryOutput,
    enrichmentResult: EnrichmentOutput,
  ): string {
    return [
      'Output a push decision based on the following analysis. Output structured JSON only.',
      `News Title: ${summaryResult.title}`,
      `Summary: ${summaryResult.summary}`,
      `Category: ${classificationResult.category}`,
      `Tags: ${classificationResult.tags.join(', ')}`,
      `Hot Score: ${hotspotResult.score}`,
      `Priority: ${hotspotResult.priority}`,
      `Trend: ${hotspotResult.trend}`,
      `Impact: ${enrichmentResult.impact.slice(0, 200)}`,
      `Model: ${buildModelLabel(ctx.model_name)}`,
      `Channels: ${(ctx.config.push_channels as string[]) || 'webhook,email'}`,
    ].join('\n');
  }

  async run(
    ctx: AgentExecutionContext,
    hotspotResult: HotspotOutput,
    classificationResult: ClassificationOutput,
    summaryResult: SummaryOutput,
    enrichmentResult: EnrichmentOutput,
  ): Promise<PushPlanOutput> {
    const aiResult = await this.runtime.callStructured<PushPlanOutput>({
      agentName: 'PushPlannerAgent',
      systemPrompt: ctx.prompt_bundle.agent_push_planner_prompt || '',
      userPrompt: this.buildUserPrompt(
        ctx,
        hotspotResult,
        classificationResult,
        summaryResult,
        enrichmentResult,
      ),
      schema: PushPlanSchema,
    });

    if (aiResult) {
      this.logger.debug(`PushPlannerAgent AI result: push_now=${aiResult.push_now}, priority=${aiResult.priority}`);
      return aiResult;
    }

    return this.fallback(hotspotResult, classificationResult, summaryResult, ctx);
  }

  private fallback(
    hotspotResult: HotspotOutput,
    classificationResult: ClassificationOutput,
    summaryResult: SummaryOutput,
    ctx: AgentExecutionContext,
  ): PushPlanOutput {
    const priority = hotspotResult.priority;
    const channels = recommendChannels(priority);
    const configuredChannels = ctx.config.push_channels as string[] | undefined;

    const now = new Date();
    const pushNow = priority === 'HIGH';
    const plannedAt = pushNow
      ? now.toISOString()
      : new Date(now.getTime() + 6 * 60 * 60 * 1000).toISOString();

    return {
      push_now: pushNow,
      priority,
      push_type: 'HIGHLIGHT',
      channels: mergeChannels(channels, configuredChannels),
      planned_at: plannedAt,
      reason:
        priority === 'HIGH'
          ? `High hot content (${hotspotResult.score}), recommend immediate push`
          : `Medium hot content (${hotspotResult.score}), scheduled for later`,
      title: summaryResult.title,
    };
  }
}
